packages/swingset-liveslots/test/gc-and-finalize.js

```diff
@@ -0,0 +1,91 @@
+/* global setImmediate */
+
+/* A note on our GC terminology:
+ *
+ * We define four states for any JS object to be in:
+ *
+ * REACHABLE: There exists a path from some root (live export or top-level
+ * global) to this object, making it ineligible for collection. Userspace vat
+ * code has a strong reference to it (and userspace is not given access to
+ * WeakRef, so it has no weak reference that might be used to get access).
+ *
+ * UNREACHABLE: There is no strong reference from a root to the object.
+ * Userspace vat code has no means to access the object, although liveslots
+ * might (via a WeakRef). The object is eligible for collection, but that
+ * collection has not yet happened. The liveslots WeakRef is still alive: if
+ * it were to call `.deref()`, it would get the object.
+ *
+ * COLLECTED: The JS engine has performed enough GC to notice the
+ * unreachability of the object, and has collected it. The liveslots WeakRef
+ * is dead: `wr.deref() === undefined`. Neither liveslots nor userspace has
+ * any way to reach the object, and never will again. A finalizer callback
+ * has been queued, but not yet executed.
+ *
+ * FINALIZED: The JS engine has run the finalizer callback. After this point,
```

Comment:
yes to both, the main distinction between COLLECTED and FINALIZED is whether or not the callback has been called, since there's this interim state where the object is completely unreachable, and if you poll the weakref (call `wr.deref()`) you'll get `undefined`, but the callback is still waiting to happen.

@mhofman has pointed out that we shouldn't be polling a weakref, since if the object is in the UNREACHABLE state, the act of polling will inhibit actual collection for a while longer. We do the poll because an object might get revived (by e.g. a new delivery that re-imports a previously-UNREACHABLE Presence), but the finalizer callback is still going to happen, so we need to make sure the callback doesn't accidentally clean up after an object that isn't really dead. @mhofman 's recommendation is to cancel the finalizer entry (unregister it) upon re-import, rather than have the callback re-check the weakref. But there are details that we need to work out.

Reply:
Thanks for the details. Wrt this PR, please clarify "After this point" because it reads as if "after the point the callback has been called", which confuses the rest of the paragraph.

Based on your confirmation, I think this would suffice:

> FINALIZED: The JS engine has run the finalizer callback. *Once the callback completes,*

Reply:
done, in both copies